File: server1.cpp
#include <winsock2.h>
#include <windows.h>
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <mutex>

#pragma comment(lib, "ws2_32.lib")

std::mutex data_mutex;

// Oxirgi kelgan ma'lumotlar (yagona obyekt)
std::string lastJsonData = R"({ "status": "no data yet" })";

// POST JSON ma'lumotni olish
std::string extractPostData(const std::string& request) {
    size_t pos = request.find("\r\n\r\n");
    if (pos != std::string::npos) {
        return request.substr(pos + 4);
    }
    return "";
}

void handleClient(SOCKET client_socket) {
    char buffer[2048];
    int recv_size = recv(client_socket, buffer, sizeof(buffer) - 1, 0);
    if (recv_size == SOCKET_ERROR) {
        closesocket(client_socket);
        return;
    }

    buffer[recv_size] = '\0';
    std::string request(buffer);

    std::string response_body;
    std::string status = "200 OK";

    if (request.find("POST /upload") != std::string::npos) {
        std::string json = extractPostData(request);

        {
            std::lock_guard<std::mutex> lock(data_mutex);
            lastJsonData = json;
        }

        response_body = R"({"status": "received"})";

    } else if (request.find("GET /data") != std::string::npos) {
        {
            std::lock_guard<std::mutex> lock(data_mutex);
            response_body = lastJsonData;
        }
    } else {
        status = "404 Not Found";
        response_body = R"({"error": "invalid endpoint"})";
    }

    std::ostringstream response;
    response << "HTTP/1.1 " << status << "\r\n"
             << "Content-Type: application/json\r\n"
             << "Access-Control-Allow-Origin: *\r\n"
             << "Content-Length: " << response_body.size() << "\r\n"
             << "Connection: close\r\n\r\n"
             << response_body;

    send(client_socket, response.str().c_str(), response.str().size(), 0);
    closesocket(client_socket);
}

int main() {
    WSADATA wsa;
    SOCKET server_socket, client_socket;
    struct sockaddr_in server, client;
    int c;

    WSAStartup(MAKEWORD(2, 2), &wsa);
    server_socket = socket(AF_INET, SOCK_STREAM, 0);

    server.sin_family = AF_INET;
    server.sin_addr.s_addr = INADDR_ANY; // Agar global IP bo‘lsa, shuni qo‘ying
    server.sin_port = htons(8080); // Ehtimol bu portni routerning port forwardingida ochishingiz kerak bo‘ladi

    bind(server_socket, (struct sockaddr*)&server, sizeof(server));
    listen(server_socket, 5);

    std::cout << "Server ishga tushdi. Port: 8080\n";

    c = sizeof(struct sockaddr_in);

    while (true) {
        client_socket = accept(server_socket, (struct sockaddr*)&client, &c);
        if (client_socket != INVALID_SOCKET) {
            CreateThread(nullptr, 0, (LPTHREAD_START_ROUTINE)handleClient, (LPVOID)client_socket, 0, nullptr);
        }
    }

    closesocket(server_socket);
    WSACleanup();
    return 0;
}

File: server1.py
import json
import socket
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler

# Global o'zgaruvchi
latest_data = {"status": "no data yet"}
data_lock = threading.Lock()

# ==========================
#   HTTP JSON server qismi
# ==========================
class CarTrackerHandler(BaseHTTPRequestHandler):
    def _set_headers(self, status=200):
        self.send_response(status)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()

    def do_GET(self):
        if self.path == "/data":
            with data_lock:
                response = json.dumps(latest_data)
            self._set_headers()
            self.wfile.write(response.encode('utf-8'))
        else:
            self.send_error(404, "Not Found")

def run_http_server(port=8080):
    server_address = ('', port)
    httpd = HTTPServer(server_address, CarTrackerHandler)
    print(f"✅ HTTP server ishga tushdi: http://0.0.0.0:{port}/data")
    httpd.serve_forever()

# ==========================
#   TCP qismi (ma’lumot qabul qiladi)
# ==========================
def run_tcp_server(host='0.0.0.0', port=9000):
    global latest_data
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket:
        server_socket.bind((host, port))
        server_socket.listen(5)
        print(f"📡 TCP server ishga tushdi: {host}:{port}")

        while True:
            conn, addr = server_socket.accept()
            with conn:
                print(f"🟢 Yangi TCP ulanish: {addr}")
                data = conn.recv(4096)
                if not data:
                    continue
                try:
                    json_data = json.loads(data.decode('utf-8'))
                    with data_lock:
                        latest_data = json_data
                    print("📩 Ma’lumot qabul qilindi:", json_data)
                    conn.sendall(b'{"status": "received"}')
                except json.JSONDecodeError:
                    print("⚠️ Noto‘g‘ri JSON!")
                    conn.sendall(b'{"error": "invalid json"}')

# ==========================
#   Asosiy ishga tushirish
# ==========================
if __name__ == "__main__":
    http_thread = threading.Thread(target=run_http_server, daemon=True)
    http_thread.start()
    run_tcp_server()
